src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -91,46 +70,83 @@ public BoundLambda(SyntaxNode syntax, BoundBlock body, ImmutableArray<Diagnostic
             );
         }
 
-        public TypeSymbolWithAnnotations InferredReturnType(ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        public TypeSymbolWithAnnotations GetInferredReturnType(ref HashSet<DiagnosticInfo> useSiteDiagnostics, NullableWalker.VariableState nullableState = null)
         {
-#if DEBUG
-            Debug.Assert(_hasInferredReturnType);
-#endif
-
-            if (!_inferredReturnTypeUseSiteDiagnostics.IsNullOrEmpty())
+            if (!InferredReturnType.UseSiteDiagnostics.IsEmpty)
             {
                 if (useSiteDiagnostics == null)
                 {
                     useSiteDiagnostics = new HashSet<DiagnosticInfo>();
                 }
-
-                foreach (var info in _inferredReturnTypeUseSiteDiagnostics)
+                foreach (var info in InferredReturnType.UseSiteDiagnostics)
                 {
                     useSiteDiagnostics.Add(info);
                 }
             }
-
-            return _inferredReturnType;
+            if (nullableState == null)
+            {
+                return InferredReturnType.Type;
+            }
+            else
+            {
+                var returnTypes = ArrayBuilder<(RefKind, TypeSymbolWithAnnotations)>.GetInstance();
+                // Diagnostics from NullableWalker.Analyze can be dropped here since Analyze
+                // will be called again from NullableWalker.ApplyConversion when the
+                // BoundLambda is converted to an anonymous function.
+                var diagnostics = DiagnosticBag.GetInstance();
+                var delegateType = Type.GetDelegateType();
+                var compilation = Binder.Compilation;
+                var conversions = Binder.Conversions.WithNullability(includeNullability: true);
+                NullableWalker.Analyze(compilation, this, diagnostics, delegateInvokeMethod: delegateType?.DelegateInvokeMethod, returnTypes: returnTypes, initialState: nullableState);
+                diagnostics.Free();
+                var inferredReturnType = InferReturnType(returnTypes, compilation, conversions, delegateType, Symbol.IsAsync);
+                returnTypes.Free();
+                return inferredReturnType.Type;
+            }
         }
 
+        /// <summary>
+        /// Indicates the type of return statement with no expression. Used in InferReturnType.
+        /// </summary>
+        internal static readonly TypeSymbolWithAnnotations NoReturnExpression = TypeSymbolWithAnnotations.Create(new UnsupportedMetadataTypeSymbol());
+
```

Comment:
I would expect the `void` type to be appropriate here.

Reply:
We just need a placeholder type, ideally one that cannot be used as the type of a  `return` expression. Using `void` would require a unique instance per `Compilation` and it might cause issues when mixed with `return VoidMethod();` statements. 

---
In reply to: [193186040](https://github.com/dotnet/roslyn/pull/27283#discussion_r193186040) [](ancestors = 193186040)